### PR TITLE
Polygon drawing on map from Local DB.

### PR DESCRIPTION
**Pull request by HemanParbhakar**
Fixes #358 

Polygon drawing on map from Local DB for polygon drawing functionality into the application
Screenshot_1624362091

@atalekar @gino-m PTAL?

### CODE REVIEW

File: gnd/src/main/java/com/google/android/gnd/persistence/local/room/entity/FeatureEntity.java [hunk 1]
```diff
@@ -134,10 +168,33 @@ public static Feature toFeature(FeatureEntity featureEntity, Project project) {
...
         "No geometry data found in feature " + featureEntity.getId());
   }
 
+  public static ImmutableList<Point> stringToList(String vertices) {
+    if (vertices == null) {
+      Timber.d("vertices are null");
+      return null;
+    }
+    Gson gson = new Gson();
+    ArrayList<ArrayList<Double>> verticesArray =
```
Comment (gino-m) [inline gnd/src/main/java/com/google/android/gnd/persistence/local/room/entity/FeatureEntity.java:188]: Does `List<List<Double>>` work here?

File: gnd/src/main/java/com/google/android/gnd/persistence/local/room/entity/FeatureEntity.java [hunk 2]
```diff
@@ -134,10 +168,33 @@ public static Feature toFeature(FeatureEntity featureEntity, Project project) {
...
+        gson.fromJson(vertices, new TypeToken<ArrayList<ArrayList<Double>>>(){}.getType());
```
Comment (gino-m) [inline gnd/src/main/java/com/google/android/gnd/persistence/local/room/entity/FeatureEntity.java:189]: What happens if the vertices are empty or malformed? Is there an exception you can catch to log and return a default value?

File: gnd/src/main/java/com/google/android/gnd/persistence/local/room/entity/FeatureEntity.java [hunk 3]
```diff
@@ -134,10 +168,33 @@ public static Feature toFeature(FeatureEntity featureEntity, Project project) {
...
+    }
+
     throw new LocalDataConsistencyException(
         "No geometry data found in feature " + featureEntity.getId());
   }
 
+  public static ImmutableList<Point> stringToList(String vertices) {
+    if (vertices == null) {
+      Timber.d("vertices are null");
+      return null;
```
Comment (gino-m) [inline gnd/src/main/java/com/google/android/gnd/persistence/local/room/entity/FeatureEntity.java:185]: This method isn't declared as `@Nullable`. returning `ImmutableList.of()` might be more fault resistant.

File: gnd/src/main/java/com/google/android/gnd/persistence/local/room/entity/FeatureEntity.java [hunk 4]
```diff
@@ -134,10 +168,33 @@ public static Feature toFeature(FeatureEntity featureEntity, Project project) {
...
+    }
+    Gson gson = new Gson();
+    ArrayList<ArrayList<Double>> verticesArray =
+        gson.fromJson(vertices, new TypeToken<ArrayList<ArrayList<Double>>>(){}.getType());
+
+    return stream(verticesArray).map(vertice -> Point.newBuilder()
```
Comment (gino-m) [inline gnd/src/main/java/com/google/android/gnd/persistence/local/room/entity/FeatureEntity.java:191]: vertice -> vertex here and throughout

"vertices" is still correct for the plural form and doesn't need to be changed.

File: gnd/src/main/java/com/google/android/gnd/persistence/local/room/entity/FeatureMutationEntity.java [hunk 5]
```diff
@@ -66,13 +76,19 @@
   @Embedded
   public abstract Coordinates getNewLocation();
 
+  @CopyAnnotations
+  @Nullable
+  @ColumnInfo(name = "polygon_vertices")
+  public abstract String getPolygonVertices();
```
Comment (gino-m) [inline gnd/src/main/java/com/google/android/gnd/persistence/local/room/entity/FeatureMutationEntity.java:82]: `getNewPolygonVertices()`?

File: gnd/src/main/java/com/google/android/gnd/persistence/local/room/entity/FeatureEntity.java [hunk 6]
```diff
@@ -134,10 +168,33 @@ public static Feature toFeature(FeatureEntity featureEntity, Project project) {
...
+          PolygonFeature.builder().setVertices(stringToList(featureEntity.getPolygonVertices()));
+      fillFeature(builder, featureEntity, project);
+      return builder.build();
+    }
+
     throw new LocalDataConsistencyException(
         "No geometry data found in feature " + featureEntity.getId());
   }
 
+  public static ImmutableList<Point> stringToList(String vertices) {
```
Comment (gino-m) [inline gnd/src/main/java/com/google/android/gnd/persistence/local/room/entity/FeatureEntity.java:182]: Without context it's not clear that `stringToList()` operates on a list of vertices. How about `parsePolygonVertices()`?

File: gnd/src/main/java/com/google/android/gnd/persistence/local/room/entity/FeatureMutationEntity.java [hunk 7]
```diff
@@ -81,13 +97,44 @@ public static FeatureMutationEntity fromMutation(FeatureMutation m) {
         .build();
   }
 
+  public static String listToString(ImmutableList<Point> vertices) {
```
Comment (gino-m) [inline gnd/src/main/java/com/google/android/gnd/persistence/local/room/entity/FeatureMutationEntity.java:100]: Same comment as `stringToList()`. How about `formatPolygonVertices()`?

File: gnd/src/main/java/com/google/android/gnd/persistence/local/room/entity/FeatureEntity.java [hunk 8]
```diff
@@ -99,9 +116,24 @@ public static FeatureEntity fromMutation(FeatureMutation mutation, AuditInfo cre
             .setCreated(authInfo)
             .setLastModified(authInfo);
     mutation.getNewLocation().map(Coordinates::fromPoint).ifPresent(entity::setLocation);
+    mutation.getNewPolygonVertices().map(FeatureEntity::listToString)
+        .ifPresent(entity::setPolygonVertices);
     return entity.build();
   }
 
+  public static String listToString(ImmutableList<Point> vertices) {
+    if (vertices == null) {
```
Comment (gino-m) [inline gnd/src/main/java/com/google/android/gnd/persistence/local/room/entity/FeatureEntity.java:125]: `getNewPolygonVertices()` optional never returns a null, and `vertices` is a non-null arg (no `@Nullable` declaration), so this check can be removed.

File: gnd/src/main/java/com/google/android/gnd/persistence/local/room/entity/FeatureEntity.java [hunk 9]
```diff
@@ -99,9 +116,24 @@ public static FeatureEntity fromMutation(FeatureMutation mutation, AuditInfo cre
...
+      Timber.d("vertices are null");
+      return null;
+    }
+    Gson gson = new Gson();
+    List<List<Double>> verticesArray = stream(vertices)
+        .map(point -> ImmutableList.of(point.getLatitude(),
+            point.getLongitude())).collect(
```
Comment (gino-m) [inline gnd/src/main/java/com/google/android/gnd/persistence/local/room/entity/FeatureEntity.java:132]: Are you using `google-java-format`?

**Review by gino-m — CHANGES_REQUESTED** [review]
Hi @HemanParbhakar I started providing a review, but I realized these PRs overlap a lot. Would it be possible to send a single PR starting with just the changes to the local db? Please also be sure to give the PR a more accurate title (e.g., if the PR is focused on adding polygons to the local db, say that as opposed to "polygon drawing on map"). 

Please take some these suggestions into account while preparing that PR and be sure to include unit tests. Integration/E2E tests aren't needed. Thanks!

File: gnd/src/main/java/com/google/android/gnd/persistence/local/room/entity/FeatureEntity.java (hunk 6 above)
Comment (HemanParbhakar) [reply gnd/src/main/java/com/google/android/gnd/persistence/local/room/entity/FeatureEntity.java:182]: its converting the string to list that the reason this name was provided but parsePolygonVertices make more sense.

File: gnd/src/main/java/com/google/android/gnd/persistence/local/room/entity/FeatureEntity.java (hunk 1 above)
Comment (HemanParbhakar) [reply gnd/src/main/java/com/google/android/gnd/persistence/local/room/entity/FeatureEntity.java:188]: replaced

File: gnd/src/main/java/com/google/android/gnd/persistence/local/room/entity/FeatureEntity.java (hunk 4 above)
Comment (HemanParbhakar) [reply gnd/src/main/java/com/google/android/gnd/persistence/local/room/entity/FeatureEntity.java:191]: replaced

File: gnd/src/main/java/com/google/android/gnd/persistence/local/room/entity/FeatureMutationEntity.java (hunk 5 above)
Comment (HemanParbhakar) [reply gnd/src/main/java/com/google/android/gnd/persistence/local/room/entity/FeatureMutationEntity.java:82]: updated  -> getNewPolygonVertices

File: gnd/src/main/java/com/google/android/gnd/persistence/local/room/entity/FeatureEntity.java (hunk 6 above)
Comment (HemanParbhakar) [reply gnd/src/main/java/com/google/android/gnd/persistence/local/room/entity/FeatureEntity.java:182]: renamed -> parsePolygonVertices

**Pull request closed without merging**